File: Data Structures & Algorithms/count-connected-components/submission-0.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:

        visit = [False] * n
        adj  = {}
        for i in range(n):
            adj [i] = []

        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        def dfs(node):
            for nei in adj[node]:
                if not visit[nei]:
                    visit[nei] = True
                    dfs(nei)
        

        countComponent = 0

        for node in range(n):
            if not visit[node]:
                visit[node] = True
                dfs(node)
                countComponent +=1
        return countComponent

        

        

File: Data Structures & Algorithms/cheapest-flight-path/submission-0.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {

    vector<int> prices(n, INT_MAX);
    prices[src] = 0;
    vector<vector<pair<int, int>>> adj(n);
    //Shortest Path Faster Algorithm
    for (const auto& flight : flights){
        adj[flight[0]].emplace_back(flight[1], flight[2]);
    }

    queue<tuple<int, int, int>> q;
    //First 0 → total cost so far
    //Second 0 → current airport (src)
    //Third 0 → number of stops taken so far
    q.push({0, src, 0});


    while(!q.empty()){
        auto [cst, node, stops] = q.front();
        q.pop();

        if(stops > k){
            continue;
        }

        for(const auto& neighbor : adj[node]){
            int nei = neighbor.first, w = neighbor.second;
            cout << " nei : " <<nei << "\nw: " << w;
            int nextCost = cst + w;
            if(nextCost <  prices[nei]){
                prices[nei] = nextCost;
                q.push({nextCost, nei, stops+1});
            }
        }
    }
    return prices[dst] == INT_MAX ? -1 : prices[dst];
    }

};
